File: backend/app/services/timings_parser.py
"""
Simple restaurant timings parser.

It reads strings like:
Mon-Sun: 6:30 AM - 11 PM
Tue-Sat 12 Noon to 3 PM, 7 PM to 11 PM
Mon, Wed-Fri: 10 AM - 8 PM; Tue: Closed
"""
import argparse
import csv
import json
import re


DAYS = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

DAY_ALIASES = {
    "mon": "Mon",
    "monday": "Mon",
    "tue": "Tue",
    "tues": "Tue",
    "tuesday": "Tue",
    "wed": "Wed",
    "wednesday": "Wed",
    "thu": "Thu",
    "thur": "Thu",
    "thurs": "Thu",
    "thursday": "Thu",
    "fri": "Fri",
    "friday": "Fri",
    "sat": "Sat",
    "saturday": "Sat",
    "sun": "Sun",
    "sunday": "Sun",
}

DAY_PATTERN = (
    r"(?:mon(?:day)?|tue(?:s|sday)?|wed(?:nesday)?|"
    r"thu(?:r|rs|rsday|rday)?|fri(?:day)?|sat(?:urday)?|sun(?:day)?)"
)


def clean_text(value):
    if not value:
        return ""

    text = str(value).strip()
    for old, new in {"\u2013": "-", "\u2014": "-", "\u2212": "-", "\u00a0": " "}.items():
        text = text.replace(old, new)

    text = re.sub(r"\s+", " ", text)
    text = re.sub(r"\bto\b", "-", text, flags=re.I)
    text = re.sub(r"\bnoon\b", "12 PM", text, flags=re.I)
    text = re.sub(r"\bmidnight\b", "12 AM", text, flags=re.I)
    text = re.sub(r"(\d)(am|pm)\b", r"\1 \2", text, flags=re.I)
    return text.strip()


def normalize_day(day_text):
    key = re.sub(r"[^a-z]", "", day_text.lower())
    return DAY_ALIASES.get(key)


def expand_day_token(token):
    range_match = re.match(
        rf"^\s*({DAY_PATTERN})\s*-\s*({DAY_PATTERN})\s*$",
        token.strip(),
        flags=re.I,
    )
    if range_match:
        start = normalize_day(range_match.group(1))
        end = normalize_day(range_match.group(2))
        if not start or not end:
            return []

        start_index = DAYS.index(start)
        end_index = DAYS.index(end)
        if start_index <= end_index:
            return DAYS[start_index : end_index + 1]
        return DAYS[start_index:] + DAYS[: end_index + 1]

    found = []
    for day in re.findall(DAY_PATTERN, token, flags=re.I):
        normalized = normalize_day(day)
        if normalized and normalized not in found:
            found.append(normalized)
    return found


def expand_days(days_text):
    days = []
    for token in re.split(r"\s*,\s*|\s*&\s*|\s+and\s+", days_text, flags=re.I):
        for day in expand_day_token(token):
            if day not in days:
                days.append(day)
    return days


def normalize_time_piece(piece):
    piece = piece.strip().upper().replace(".", "")
    piece = re.sub(r"\s+", " ", piece)
    match = re.match(r"^(\d{1,2})(?::(\d{2}))?\s*(AM|PM)$", piece)
    if not match:
        return piece

    hour = int(match.group(1))
    minute = match.group(2)
    suffix = match.group(3)
    if minute and minute != "00":
        return f"{hour}:{minute} {suffix}"
    return f"{hour} {suffix}"


def normalize_times(times_text):
    text = clean_text(times_text)
    if not text:
        return ""
    if re.search(r"\b(closed|close|holiday|not open)\b", text, flags=re.I):
        return "Closed"
    if re.search(r"\b(24\s*hours|open\s*24|24/7)\b", text, flags=re.I):
        return "Open 24 hours"

    time_re = r"\d{1,2}(?::\d{2})?\s*(?:AM|PM)"
    ranges = []
    for start, end in re.findall(rf"({time_re})\s*-\s*({time_re})", text, flags=re.I):
        ranges.append(f"{normalize_time_piece(start)} - {normalize_time_piece(end)}")

    return ", ".join(ranges) if ranges else text


def split_timing_sections(text):
    pattern = re.compile(
        rf"((?:{DAY_PATTERN})(?:\s*-\s*(?:{DAY_PATTERN}))?"
        rf"(?:\s*,\s*(?:{DAY_PATTERN})(?:\s*-\s*(?:{DAY_PATTERN}))?)*)"
        rf"\s*:?\s*",
        flags=re.I,
    )
    matches = list(pattern.finditer(text))
    sections = []

    for index, match in enumerate(matches):
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(text)
        times_text = text[start:end].strip(" ;,")
        if times_text:
            sections.append((match.group(1), times_text))

    return sections


def parse_timings(raw_timings):
    schedule = {day: "Closed" for day in DAYS}
    text = clean_text(raw_timings)
    if not text:
        return schedule

    sections = split_timing_sections(text) or [("Mon-Sun", text)]
    for days_text, times_text in sections:
        times = normalize_times(times_text)
        for day in expand_days(days_text):
            schedule[day] = times

    return schedule


def convert_csv(input_path, output_path, timings_column="timings"):
    with open(input_path, newline="", encoding="utf-8-sig") as input_file:
        reader = csv.DictReader(input_file)
        fieldnames = list(reader.fieldnames or [])
        actual_column = next(
            (name for name in fieldnames if name.lower() == timings_column.lower()),
            timings_column,
        )
        output_column = "structured_timings"
        if output_column not in fieldnames:
            fieldnames.append(output_column)

        rows = []
        for row in reader:
            row[output_column] = json.dumps(
                parse_timings(row.get(actual_column, "")),
                ensure_ascii=False,
            )
            rows.append(row)

    with open(output_path, "w", newline="", encoding="utf-8") as output_file:
        writer = csv.DictWriter(output_file, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def main():
    parser = argparse.ArgumentParser(description="Parse restaurant timings in a CSV file.")
    parser.add_argument("input_csv", help="Path to the input CSV file")
    parser.add_argument("output_csv", help="Path to write the output CSV file")
    parser.add_argument("--column", default="timings", help="Timings column name")
    args = parser.parse_args()

    convert_csv(args.input_csv, args.output_csv, args.column)
    print(f"Done. Wrote parsed timings to {args.output_csv}")


if __name__ == "__main__":
    main()

File: frontend/src/utils/timingsParser.js
const DAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun']

const DAY_ALIASES = {
  mon: 'Mon',
  monday: 'Mon',
  tue: 'Tue',
  tues: 'Tue',
  tuesday: 'Tue',
  wed: 'Wed',
  wednesday: 'Wed',
  thu: 'Thu',
  thur: 'Thu',
  thurs: 'Thu',
  thursday: 'Thu',
  fri: 'Fri',
  friday: 'Fri',
  sat: 'Sat',
  saturday: 'Sat',
  sun: 'Sun',
  sunday: 'Sun',
}

const DAY_PATTERN = '(?:mon(?:day)?|tue(?:s|sday)?|wed(?:nesday)?|thu(?:r|rs|rsday|rday)?|fri(?:day)?|sat(?:urday)?|sun(?:day)?)'

function cleanText(value = '') {
  return String(value)
    .replace(/[–—−]/g, '-')
    .replace(/\u00a0/g, ' ')
    .replace(/\bto\b/gi, '-')
    .replace(/\bnoon\b/gi, '12 PM')
    .replace(/\bmidnight\b/gi, '12 AM')
    .replace(/(\d)(am|pm)\b/gi, '$1 $2')
    .replace(/\s+/g, ' ')
    .trim()
}

function emptySchedule() {
  return DAYS.reduce((schedule, day) => {
    schedule[day] = 'Closed'
    return schedule
  }, {})
}

function normalizeDay(dayText) {
  return DAY_ALIASES[dayText.toLowerCase().replace(/[^a-z]/g, '')]
}

function expandDayToken(token) {
  const rangeRe = new RegExp(`^\\s*(${DAY_PATTERN})\\s*-\\s*(${DAY_PATTERN})\\s*$`, 'i')
  const rangeMatch = token.match(rangeRe)
  if (rangeMatch) {
    const start = normalizeDay(rangeMatch[1])
    const end = normalizeDay(rangeMatch[2])
    if (!start || !end) return []

    const startIndex = DAYS.indexOf(start)
    const endIndex = DAYS.indexOf(end)
    if (startIndex <= endIndex) return DAYS.slice(startIndex, endIndex + 1)
    return [...DAYS.slice(startIndex), ...DAYS.slice(0, endIndex + 1)]
  }

  const days = []
  const dayRe = new RegExp(DAY_PATTERN, 'gi')
  for (const match of token.matchAll(dayRe)) {
    const day = normalizeDay(match[0])
    if (day && !days.includes(day)) days.push(day)
  }
  return days
}

function expandDays(daysText) {
  const days = []
  daysText.split(/\s*,\s*|\s*&\s*|\s+and\s+/i).forEach(token => {
    expandDayToken(token).forEach(day => {
      if (!days.includes(day)) days.push(day)
    })
  })
  return days
}

function normalizeTimePiece(piece) {
  const match = piece.trim().toUpperCase().replace(/\./g, '').match(/^(\d{1,2})(?::(\d{2}))?\s*(AM|PM)$/)
  if (!match) return piece.trim()

  const hour = Number(match[1])
  const minute = match[2]
  const suffix = match[3]
  return minute && minute !== '00' ? `${hour}:${minute} ${suffix}` : `${hour} ${suffix}`
}

function normalizeTimes(timesText) {
  const text = cleanText(timesText)
  if (!text) return ''
  if (/\b(closed|close|holiday|not open)\b/i.test(text)) return 'Closed'
  if (/\b(24\s*hours|open\s*24|24\/7)\b/i.test(text)) return 'Open 24 hours'

  const timeRe = '\\d{1,2}(?::\\d{2})?\\s*(?:AM|PM)'
  const rangeRe = new RegExp(`(${timeRe})\\s*-\\s*(${timeRe})`, 'gi')
  const ranges = []

  for (const match of text.matchAll(rangeRe)) {
    ranges.push(`${normalizeTimePiece(match[1])} - ${normalizeTimePiece(match[2])}`)
  }
  return ranges.length ? ranges.join(', ') : text
}

function splitTimingSections(text) {
  const sectionRe = new RegExp(`((?:${DAY_PATTERN})(?:\\s*-\\s*(?:${DAY_PATTERN}))?(?:\\s*,\\s*(?:${DAY_PATTERN})(?:\\s*-\\s*(?:${DAY_PATTERN}))?)*)\\s*:?\\s*`, 'gi')
  const matches = [...text.matchAll(sectionRe)]

  return matches.map((match, index) => {
    const start = match.index + match[0].length
    const end = matches[index + 1]?.index ?? text.length
    return [match[1], text.slice(start, end).replace(/^[\s;,]+|[\s;,]+$/g, '')]
  }).filter(([, timesText]) => timesText)
}

export function parseTimings(rawTimings) {
  if (rawTimings && typeof rawTimings === 'object') return { ...emptySchedule(), ...rawTimings }

  const schedule = emptySchedule()
  const text = cleanText(rawTimings)
  if (!text) return schedule

  const sections = splitTimingSections(text)
  const usableSections = sections.length ? sections : [['Mon-Sun', text]]

  usableSections.forEach(([daysText, timesText]) => {
    const times = normalizeTimes(timesText)
    expandDays(daysText).forEach(day => {
      schedule[day] = times
    })
  })
  return schedule
}

export function getTodayKey(date = new Date()) {
  return DAYS[(date.getDay() + 6) % 7]
}

export function timeToMinutes(timeText) {
  const match = timeText.trim().toUpperCase().match(/^(\d{1,2})(?::(\d{2}))?\s*(AM|PM)$/)
  if (!match) return null

  let hour = Number(match[1])
  const minute = Number(match[2] || 0)
  const suffix = match[3]
  if (suffix === 'AM' && hour === 12) hour = 0
  if (suffix === 'PM' && hour !== 12) hour += 12
  return hour * 60 + minute
}

export function isOpenAt(dayText, date = new Date()) {
  if (!dayText || /^closed$/i.test(dayText)) return false
  if (/24\s*hours/i.test(dayText)) return true

  const now = date.getHours() * 60 + date.getMinutes()
  const timeRe = '\\d{1,2}(?::\\d{2})?\\s*(?:AM|PM)'
  const rangeRe = new RegExp(`(${timeRe})\\s*-\\s*(${timeRe})`, 'gi')

  for (const match of dayText.matchAll(rangeRe)) {
    const start = timeToMinutes(match[1])
    const end = timeToMinutes(match[2])
    if (start == null || end == null) continue
    if (end <= start && (now >= start || now <= end)) return true
    if (end > start && now >= start && now <= end) return true
  }
  return false
}

export { DAYS }
